mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSourceSchemaReader.java

```diff
@@ -61,27 +65,29 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
 
   /**
    * Whether the corresponding column should be converted to CDAP Datetime Logical Type
+   * SQL Server data type datetime and datetime2 will be converted to CDAP Datetime Logical Type
    * @param metadata result set metadata
    * @param index index of the column
    * @return whether the corresponding column should be converted to CDAP Datetime Logical Type
    * @throws SQLException
    */
   public static boolean shouldConvertToDatetime(ResultSetMetaData metadata, int index) throws SQLException {
-    // datetimeoffset will have type DATETIME_OFFSET_TYPE
-    // datetime and datetime2 will have type Types.TIMESTAMP
+    // DateTimeOffset will have type DATETIME_OFFSET_TYPE
     // cannot decide based on sql type
     String columnTypeName = metadata.getColumnTypeName(index);
-    return shouldConvertToDatetime(columnTypeName);
-  }
+    int columnSqlType = metadata.getColumnType(index);
 
-  /**
-   * Whether the corresponding data type should be converted to CDAP Datetime Logical Type
-   * SQL Server data type datetime, datetime2 and datetimeoffset will be converted to CDAP Datetime Logical Type
-   * @param typeName the data type name
-   * @return Whether the corresponding data type should be converted to CDAP Datetime Logical Type
-   */
-  public static boolean shouldConvertToDatetime(String typeName) {
-    return typeName.startsWith(DATETIME_TYPE_PREFIX);
+    switch (columnSqlType) {
+      case Types.TIMESTAMP:
+        // SmallDateTime will also be mapped to Types.DATETIME
+        return true;
+      case DATETIME_OFFSET_TYPE:
+        // DateTimeOffset should map to Types.TIMESTAMP_MICROS
+        return false;
+      default:
+        // DateTime and DateTime2 will be converted to CDAP DateTime
+        return columnTypeName.startsWith(DATETIME_TYPE_PREFIX);
+    }
```

Comment:
You can refactor this, to check only for Types.TIMESTAMP and default case.

Reply:
done